### DISCUSSION TITLE
Language integrated SQL

### DISCUSSION THREAD

**Discussion by FrenkR** [dotnet/roslyn#68763] [Ideas]
Hello,
I am quite new to "roslyn" compiler architecture and what I currently understand, is, that it is a very powerful tool. And that it is possible to somehow extend any implemented language with "custom" contextual keyword.
Does anybody have an idea where to start, to implement simple "SQL integrated query" support for regular SQLs (select, types of join, aggregates, etc. ), that would be written almost similar that are inside SQL with defined single class per table.
for example(select at "end" is by intent - similar way as linq):
```
from orderlines a // orderlines class
 left join customer b on b.cust == a.cust // standard left join to customer class
left join orders c on c.orderid == a.orderId 
where  c.orderStatus in [OrderStatus.Open, OrderStatus.Approved] // "array of possible values
  and b.customerDescription like "%" + m_search + "%" // variable m_search

select _item = a.item, _order = c // varibale _item is mapped to a.item, variable _order is mapped to "c" that is returned by "join"
selectdo
  // cursor for each returned record
selectempty
  // no rows returned - 
endselect  // end cursor

```

**Comment by CyrusNajmabadi** [comment]
You'd first have to fork the repo.  We would not accept the changes you're describing into Roslyn itself (as this would no longer be c#), but you're free to fork and take the language wherever you want to. 

Once forked, I'd recommend looking into LanguageParser.cs and see how we parse actual linq queries. You can then start extending that however you want. Once you extend the syntax, you'll then figure out how to 'bind' it (give it semantic naming y), and finally how to 'emit' it (produce the IL that actually executes in the clr at runtime).

**Reply by FrenkR** [reply]
is fork needed? or is there any other place to extend language itself? I mean in a way of additional "dll", with sort of hook when parsing syntax and hence some "config" option during compiling/parsing?

if it will be fork, then I believe it means I get new "csc", right?

**Reply by CyrusNajmabadi** [reply]
> is fork needed?

Yes.

> or is there any other place to extend language itself? I mean in a way of additional "dll", with sort of hook when parsing syntax and hence some "config" option during compiling/parsing?

No, this is explicitly *not* supported.  The roslyn from this repo exists only to compile C# and VB and nothing else.

> if it will be fork, then I believe it means I get new "csc", right?

Correct. :)

**Reply by FrenkR** [reply]
perfect, thank you.
